dev-docs/007-transforms-rfc.md

```diff
@@ -0,0 +1,99 @@
+- Start Date: 2019-09-05
+- RFC PR: 
+
+# Summary
+
+It is desirable to enable arbitrary transform support common in robotic systems. While XVIZ models some explicitly
+with our coordinate types IDENTITY, GEOGRAPHIC, and VEHICLE_RELATIVE, the need for client controlled frames of reference
+will make it easier to interop with other systems as well as match a familiar modeling automous system data.
+
+# Motivation
+
+Some PERSISTENT data, such as *maps* and *point clouds*, could be encoded in various ways such as vehicle relative.
+If they are semantically fixed, but the reference point changes, there is no way to efficiently update them without
+sending all the points again.
+
+Ideally, the only data that is required is the transform to adjust how the points are transformed.
+
+Extending this to a general robotic systems convention, the need to describe parent/child relationships
+is a common model for robotic systems.
+
+XVIZ is not focused on articulated robotic platforms but support for the convention common when modeling autonomous data
+would be benefitial for XVIZ and it's users.
+
+# Proposal
+
+### POSE schema
+ - Add a *parent* field to the pose that is a string that my reference another pose stream
+   If present within the data, it would be used to construct the final transform matrix for the data
+ - Poses may reference any other pose as a *parent* across all state_update message types. Meaning a PERSISTENT
+   pose could be a parent of a *COMPLETE* or *INCREMENTAL* pose.
+ - The pose stream names must be unique.
+ - Remove timestamp as required field on the pose schema. The state_update defines the timestamp for the message already
+   and the pose entry was just a carry over from xviz v1
+
+### PRIMITIVE schema
+ - add a *pose* field to the primitive_base schema, making the attribute accessible to all primitives
+
+### STREAM METADATA schema
+ - Add a *pose* field, as a string referencing a POSE stream
+
+# Detailed Design
+
+The *pose* stream for a primitive may be in the metadata or on the individual elements within the pose
+It may be assumed that all elements within a stream share the same *pose*, and only first entry must be
+checked.
+
+The reasons for this are described below, but briefly it is to avoid having to scan and batch the elements
+when rendering.  If elements have a different transform they should be placed in a different layer.
+
+## Streetscape.gl implementation details
+
+TODO - quick check with xiaoji that this should not be a problem. The stream synchronizer has all the data
+at the time of render so we can query the first element or metadata and fetch out the proper transform.
+
+We may want to build a transform cache if necessary, but that is a performance implmentation detail.
+
+## Other considerations
+
+### rendering consideration
+
+  in deck.gl, used by streetscape.gl, the layers are separated by geometry type and we render all data
```

Comment:
I would expect at least the major external stream, such as the *maps* and *point clouds* discussed in the Motivation section, to be rendered in their own deck.gl layers and thus have their own `Layer.modelMatrix`, i.e. no need for per-object transforms.

This is also good practice since these data sets do not change from frame to frame, kept in a separate layer deck.gl will only process the data once and the app just feeds in new composite transformation matrix (if needed) to each layer every frame.

Reply:
Correct.

It was unclear if supporting per-object transforms would be required.  And after talking to you and Xiaoji, it is obvious it is not something XVIZ & Streetscape.gl *should* encourage. This section was to call out that consideration and collect the reasons why it is not a good feature.

We can provide a transform hierarchy but it could be abused and result in poor performance .